### PR TITLE
Multithreading via C++ thread pool of clients

### PR DESCRIPTION
**Pull request by kentslaney**
I've implemented the C++ thread workers to be FILO under the assumption that reusing a single client is preferable when possible. The current implementation needs C++17 or above to parallelize some of the client setup ([`std::execution`](https://en.cppreference.com/w/cpp/header/execution)). Building still fails with the required version, but I figured I'd open a draft PR anyway in case there's feedback. I'd especially appreciate advice about a good testing setup for the concurrent code along with any opinions about language bindings.

### CODE REVIEW

**Comment by lexdene** [comment]
World you please show a benchmark about the performance improvement your modifications have?
If there is no obvious improvement, I'd like to do the multithreading stuffs in Python code rather than inside the code of libmc.

**Comment by kentslaney** [comment]
The benchmark results can be found [here](https://github.com/douban/libmc/actions/runs/7681168674/job/20934069824#step:7:29). The naive implementation of a python thread pool comes with the problem of an arbitrarily bad worst case since the unlock order is undefined, which you can see in the standard deviation of `py thread pool`. A [thread mapped pool](https://sendapatch.se/projects/pylibmc/pooling.html#thread-mapped-pooling) handles large memory request spikes poorly, space which can be used for a local cache. It can also put the caching server under high load since the number of memcached client connections is tied to the number of HTTP response handlers. The python equivalent of the C++ implementation, `py FIFO thread pool` should solve both problems. Compared to C++, the python version has a higher average response time (~1.5x) and standard deviation (~2x), but shouldn't suffer from the same scaling problems as `py thread pool`.

The same C++ implementation can be applied to Go as well, but from a maintenance standpoint, C++ is 250+ lines compared to python's 50. Given Go doesn't have a GIL, the response time is probably going to be significantly better than 1.5x C++. Results including gevent are still pending [bug fixes on an update to greenify](https://github.com/kentslaney/greenify/tree/cond-var).

@lexdene I'm not sure what portion of server time is spent on cache I/O or if 1.5x is worth the maintenance. Let me know, since it changes whether I work on gevent hooks or moving thread support to python next.

**Comment by lexdene** [comment]
I think any improvement is welcome unless it brings breaking changes.
After looking over your code, I see that a new class with thread is introduced and the original class remains the same as before, which lgtm.

**Comment by kentslaney** [comment]
Alright, I think this is ready to be reviewed/merged. I also rewrote some parts of the README unrelated to this PR for the sake of clarity.

I'll start tracking gevent support for threaded clients in a [greenify PR](https://github.com/douban/greenify/pull/27). Ideally that shouldn't affect the implementation here.

**Comment by kentslaney** [comment]
It looks like gomemcache is [already thread safe](https://pkg.go.dev/github.com/bradfitz/gomemcache/memcache#Client), so I think this PR is done.

**Comment by lexdene** [comment]
I think the changes under `.github/workflows` directory are not related.
Would you please revert all changes under this directory?

I do not agree with some changes of readme.
Additionally, I do not want to discuss much about readme here which I think is not related.
Would you please revert all changes in readme and submit another pull request which only contains changes of it so that we can discuss specifically and pointedly.

What is `misc/aliases` used for?
I have not found any code referencing this file.
If it is possible to delete `misc/aliases`,
`misc/git/debug.patch` and `misc/git/pre-commit` which are only referenced by `misc/aliases` can also be deleted.

Changes about `greenify` are also not related I think.

Sorry for replying so slowly, but it is really not easy to look over such a big pull request.
I will be happier if it is splitted into several small ones.

**Comment by kentslaney** [comment]
I think I've now removed all the parts of the pull request that aren't necessary for the `ThreadedClient` interface in Python.

Quick summary of the parts removed:
- `.github/workflows/golang.yml` and `.github/workflows/python.yml`: changes for [nektos/act](https://github.com/nektos/act)
- `.github/workflows/manual.yml`: shorter, manual tests for faster iteration
- ` .gitignore`: cygdb (cython debugger) uses `/cython_debug`
- `Makefile`: `make clean`
- `README.rst`: discussed above
- `include/Common.h`, `src/Common.cpp`, and `src/Connection.cpp`: better naming for some of the things in #120 
- `misc/aliases`: executable with shortcuts for commands I regularly used during development
- `misc/git/*`: created git hooks that automatically un-staged the `setup.py` changes I needed for debugging Cython
- `misc/memcached_server`: logs all port traffic for debugging reasons
- `misc/runbench.py`: raises an exception for tests with incorrect output instead of potentially clogging stdout
- `setup.py`: actually I don't remember why this one was there, but it [removes some process cleanup steps apparently](https://docs.python.org/3/library/os.html#os._exit)

I'm currently expecting to make separate PRs to add back the changes in:
- `README.rst`
- `include/Common.h`, `src/Common.cpp`, and `src/Connection.cpp`

Here's some ideas on parts that can further split up the changes, along with what I see as the benefits and reasons not to. 
- benchmarking changes needed for threading
  - pros
    - allows the functional parts to be merged with less code to review
  - cons
    - mostly just moves the changes in `runbench.py` to a different PR
    - removes an integration test for the threaded code
- separate the C threading interface from the python one
  - pros
    - fairly even split, significant reduction in largest PR size
  - cons
    - they're already mostly organized by file type
    - wouldn't clarify the changes in either
    - the python interface, most likely to be user facing, has the same critical path

Please let me know if there's any changes that would make reviewing easier.

File: src/ClientPool.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  duplicate_strings(hosts, n, m_hosts_data, m_hosts);
+  duplicate_strings(aliases, n, m_aliases_data, m_aliases);
+
+  m_ports.resize(n);
+  std::copy(ports, ports + n, m_ports.begin());
+
+  std::atomic<int> rv = 0;
+  std::lock_guard<std::mutex> updating(m_fifo_access);
+  std::for_each(irange(0), irange(m_clients.size()),
+  //std::for_each(std::execution::par_unseq, irange(0), irange(m_clients.size()),
```
Comment (kentslaney) [inline src/ClientPool.cpp:69]: The loops for initializing and growing the client pool (doubles in size each time it's overrun) were written using `for_each` in order to support parallelized initialization of the client objects, but at the scales I tested it at, the parallelization didn't make a difference, thus the commented out execution policy.

File: misc/runbench.py [hunk 2]
```diff
@@ -203,9 +238,113 @@ def make_pylibmc_client(servers, **kw):
...
+
+
+class BenchmarkThreadedClient(libmc.ThreadedClient):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.config(libmc.MC_INITIAL_CLIENTS, POOL_SIZE)
+        self.config(libmc.MC_MAX_CLIENTS, POOL_SIZE)
+
+
+class FIFOThreadPool(ThreadPool):
```
Comment (kentslaney) [inline misc/runbench.py:311]: This class is analogous to the C++ `LockPool` class and might be an easier way to understand the overall control flow

**Review by lexdene — APPROVED** [review]

**Comment by kentslaney** [comment]
:tada: :tada: :tada:

**Pull request merged**